src/test/java/com/android/volley/toolbox/DiskBasedAsyncCacheTest.java

```diff
@@ -0,0 +1,449 @@
+@Config(sdk = 26)
+public class DiskBasedAsyncCacheTest {
+
+    private static final int MAX_SIZE = 1024 * 1024;
+
+    private DiskBasedAsyncCache cache;
+
+    private AsyncCache.OnCompleteCallback emptyCallback;
+
+    @Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Rule public ExpectedException exception = ExpectedException.none();
+
+    @Before
+    public void setup() throws IOException {
+        emptyCallback =
+                new AsyncCache.OnCompleteCallback() {
+                    @Override
+                    public void onComplete() {
+                        // do nothing
+                    }
+                };
+        // Initialize empty cache
+        cache = new DiskBasedAsyncCache(temporaryFolder.getRoot(), MAX_SIZE);
+        cache.initialize(emptyCallback);
+    }
+
+    @After
+    public void teardown() {
+        cache = null;
+    }
+
+    @Test
+    public void testEmptyInitialize() throws ExecutionException, InterruptedException {
+        assertNull(getEntry("key").get());
+    }
+
+    @Test
+    public void testPutGetZeroBytes() throws ExecutionException, InterruptedException {
+        final Cache.Entry entry = new Cache.Entry();
+
+        entry.data = new byte[0];
+        entry.serverDate = 1234567L;
+        entry.lastModified = 13572468L;
+        entry.ttl = 9876543L;
+        entry.softTtl = 8765432L;
+        entry.etag = "etag";
+        entry.responseHeaders = new HashMap<>();
+        entry.responseHeaders.put("fruit", "banana");
+        entry.responseHeaders.put("color", "yellow");
+
+        putEntry("my-magical-key", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("my-magical-key").get(), entry);
+        assertThat(getEntry("unknown-key").get(), is(nullValue()));
+    }
+
+    @Test
+    public void testTooLargeEntry() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("oversize") + 1);
+        putEntry("oversize", entry).get();
+
+        assertThat(getEntry("oversize").get(), is(nullValue()));
+    }
+
+    @Test
+    public void testMaxSizeEntry() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("maxsize") - 1);
+        putEntry("maxsize", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("maxsize").get(), entry);
+    }
+
+    @Test
+    public void testTrimAtThreshold() throws ExecutionException, InterruptedException {
+        // Start with the largest possible entry.
+        Cache.Entry entry = randomData(MAX_SIZE - getEntrySizeOnDisk("maxsize") - 1);
+        putEntry("maxsize", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("maxsize").get(), entry);
+
+        // Now any new entry should cause the first one to be cleared.
+        entry = randomData(0);
+        putEntry("bit", entry).get();
+
+        assertThat(getEntry("goodsize").get(), is(nullValue()));
```

Comment:
I'm a little confused by this one. Shouldn't it be checking the "maxsize" entry?

Reply:
That's what the test in DiskBasedCache did, but I think you're right that it is supposed to check the maxsize entry.

Reply:
@sophiahaim I agree. Should be maxsize. Slipped through unnoticed in https://github.com/google/volley/pull/196